dbms/src/Interpreters/Context.cpp

```diff
@@ -285,6 +290,33 @@ struct ContextShared
         background_pool.reset();
         schedule_pool.reset();
         ddl_worker.reset();
+
+        /// Trace collector is only initialized in server program
+        if (hasTraceCollector())
+        {
+            /// Stop trace collector
+            NotifyTraceCollectorToStop();
+            trace_collector_thread.join();
+
+            /// Close trace pipe - definitely nobody needs to write there after
+            /// databases shutdown
+            trace_pipe.close();
+        }
+    }
+
+    bool hasTraceCollector()
+    {
+        return trace_collector != nullptr;
+    }
+
+    void initializeTraceCollector(std::shared_ptr<TraceLog> trace_log)
+    {
+        if (trace_log == nullptr)
+            return;
+
+        trace_pipe.open();
+        trace_collector.reset(new TraceCollector(trace_log));
```

Comment:
Why don't merge all three in TraceCollector constructor.
P.S. `std::make_unique` instead of `new`.